src/Microsoft.VisualStudio.Composition/Configuration/Formatters/MetadataDictionaryFormatter.cs

```diff
@@ -0,0 +1,363 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.Composition.Formatter
+{
+    using System.Collections.Immutable;
+    using System.Globalization;
+    using System.Reflection;
+    using MessagePack;
+    using MessagePack.Formatters;
+    using Microsoft.VisualStudio.Composition.Reflection;
+
+#pragma warning disable CS8604 // Possible null reference argument.
+
+    internal class MetadataDictionaryFormatter : IMessagePackFormatter<IReadOnlyDictionary<string, object?>>
+    {
+        public static readonly MetadataDictionaryFormatter Instance = new();
+
+        private MetadataDictionaryFormatter()
+        {
+        }
+
+        /// <inheritdoc/>
+        public void Serialize(ref MessagePackWriter writer, IReadOnlyDictionary<string, object?> value, MessagePackSerializerOptions options)
+        {
+            writer.Write(value.Count);
+
+            // Special case certain values to avoid defeating lazy load later. Check out the
+            // ReadMetadata below, how it wraps the return value.
+            IReadOnlyDictionary<string, object?> serializedMetadata;
+
+            // Unwrap the metadata if its an instance of LazyMetaDataWrapper, the wrapper may end up
+            // implicitly resolving TypeRefs to Types which is undesirable.
+            value = LazyMetadataWrapper.TryUnwrap(value);
+            serializedMetadata = new LazyMetadataWrapper(value.ToImmutableDictionary(), LazyMetadataWrapper.Direction.ToSubstitutedValue, options.CompositionResolver());
+
+            foreach (KeyValuePair<string, object?> item in serializedMetadata)
+            {
+                writer.Write(item.Key);
+
+                SerializeObject(ref writer, item.Value);
+            }
+
+            void SerializeObject(ref MessagePackWriter messagePackWriter, object? value)
+            {
+                if (value is null)
+                {
+                    messagePackWriter.Write((byte)MetadataDictionaryFormatter.ObjectType.Null);
+                    return;
+                }
+
+                switch (value.GetType())
```

Comment:
If you drop `.GetType()`, then your null check cas be just a `case null:` statement.
And all the rest of your case statements get simpler too. See [this example](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYAs6A6gE4CWALgKYAUA9gEYBW9AxrQPzoA3AIYAbAK70AlPkLE5AZwDudLgAt0jYeKkykc/cS5D59dADsxIkSFkG7mOAE5GAIgtWXkgNy37+1pT0QgDWPnp+REYm6ACClJRCAJ7oQvFJNuERclBOrqkJiZ5hWf6BIcVZUaaszMwi6DV1AGqiEhkl2bmNIi3a3r4lAUGhA8QAvr4TSGNAA==).